src/Adaptor.js

```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
   };
 }
 
+/**
+ * Insert or update multiple records using ON DUPLICATE KEY
+ * @public
+ * @example
+ * upsertMany(
+ *   'users', // the DB table
+ *   [
+ *     { name: 'one', email: 'one@openfn.org' },
+ *     { name: 'two', email: 'two@openfn.org' },
+ *   ]
+ * )
+ * @constructor
+ * @param {string} table - The target table
+ * @param {array} data - An array of objects or a function that returns an array
+ * @returns {Operation}
+ */
+export function upsertMany(table, data) {
+  return function (state) {
+    /*
+    Builds and exexutes a query with syntax similar to the following 
+    (see MySQL official documentation for ON DUPLICATE KEY):
+     INSERT INTO users (name, email)
+    SELECT * FROM (
+      SELECT 'one' AS name, 'one@openfn.org' AS email UNION
+      SELECT 'two' AS name, 'two@openfn.org' AS email 
+    ) AS a
+    ON DUPLICATE KEY UPDATE name = a.name, email = a.email;
+    */
+    return new Promise(function (resolve, reject) {
+      const rows = expandReferences(data)(state); 
+
+      if (!rows || rows.length === 0) {
+        console.log('No records provided; skipping upsert.');
+        resolve(state);
+      }
+
+      const columns = Object.keys(rows[0]);
+      const columnList = columns.join(', '); 
+      
+      // 1. construct the INSERT statement
+      const insertStmt = "INSERT INTO ".concat(table, " (").concat(columnList, ")"); 
```

Comment:
Should the table name be escaped over here?

Reply:
Agree, they should be escaped.